### PR TITLE
vim-patch:8.2.{3286,3289,3293,3298,3313,3321,3328,3330,3331,3337,3354,3355,3357,3360,3369,3375}

### PR DESCRIPTION
**Pull request by seandewar**

### CODE REVIEW

**Comment by seandewar** [comment]
(did that to kickstart the CI)

**Comment by janlazo** [comment]
Needs another rebase.

**Comment by janlazo** [comment]
You shouldn't have included `test_lua.vim` patches. I intentionally omitted those so they can be ported after merging a provider to Vim's lua interface. `if_lua.c` changes are N/A to Neovim but the public api is relevant.

**Comment by janlazo** [comment]
I didn't add `test_textprop.vim`  in https://github.com/neovim/neovim/wiki/Merging-patches-from-upstream-Vim#types-of-not-applicable-vim-patches because the Vimscript api for textprops could be implemented without porting Vim's textprop implementation.

File: src/nvim/edit.c [hunk 1]
```diff
@@ -4127,6 +4127,8 @@ static int ins_compl_get_exp(pos_T *ini)
   char_u *dict = NULL;
   int dict_f = 0;
   bool set_match_pos;
+  pos_T prev_pos = { 0, 0, 0 };
+  bool looped_around = false;
```
Comment (janlazo) [inline src/nvim/edit.c:4131]: This assignment is dead. `looped_around` is set to false in the outer loop.

**Review by janlazo — CHANGES_REQUESTED** [review]
The ports to the applicable patches are fine.
Some patches considered N/A in the 1st commit are relevant because of the Vimscript **API** for lua,textprop can be ported to Neovim.

**Comment by seandewar** [comment]
Cheers for the review! I'll try and get these addressed in a bit :+1:

File: src/nvim/tag.c [hunk 2]
```diff
@@ -1467,6 +1467,7 @@ find_tags(
 
   help_save = curbuf->b_help;
   orgpat.pat = pat;
+  orgpat.regmatch.regprog = NULL;
```
Comment (janlazo) [inline src/nvim/tag.c:1470]: I wonder why Vim doesn't use `{ 0 }` (or memset) to zero-init local structs. Compiler compatiblity or optimization? Bugs from garbage struct values should be avoided.

Comment (seandewar) [reply src/nvim/tag.c:1470]: In the `funcexe_T` patches, `memset` 0 was mostly used, but there were a few cases where it wasn't.
I _think_ some later patches changed those to use `memset` too...

So I have no idea. :smile:

**Review by janlazo — APPROVED** [review]
LGTM

**Pull request merged**